### PR TITLE
Migrate to kubebuilder v3

### PR DESCRIPTION
**Pull request by maruina**
- Update to kubebuilder v3
- Update dependencies
- Add a new helper function to patch status
- Convert the api tests from ginkgo style to testing.T style

### CODE REVIEW

File: .github/workflows/test-build.yml [hunk 1]
```diff
@@ -19,15 +19,9 @@ jobs:
...
-          set -x
-          version=2.3.1
-          os=$(go env GOOS)
-          arch=$(go env GOARCH)
-          curl -Ls https://go.kubebuilder.io/dl/${version}/${os}/${arch} -o /tmp/kubebuilder_${version}.tar.gz
-          shasum -a 256 -c kubebuilder_sha256.txt
-          tar -xzf /tmp/kubebuilder_${version}.tar.gz -C /tmp/
-          sudo mv /tmp/kubebuilder_${version}_${os}_${arch} /usr/local/kubebuilder
-          export PATH=$PATH:/usr/local/kubebuilder/bin
+          curl -L -o kubebuilder https://go.kubebuilder.io/dl/latest/$(go env GOOS)/$(go env GOARCH)
```
Comment (chaudyg) [inline .github/workflows/test-build.yml:22]: Is it safe to pull the latest version?
```suggestion
          version=3.2.0
          curl -L -o kubebuilder https://go.kubebuilder.io/dl/${version}/$(go env GOOS)/$(go env GOARCH)
```

File: .github/workflows/test-build.yml [hunk 2]
```diff
@@ -15,30 +15,20 @@ jobs:
 
       - uses: actions/setup-go@v2
         with:
-          go-version: ^1.14
+          go-version: ^1.17
 
       - name: Install kubebuilder
```
Comment (Smirl) [inline .github/workflows/test-build.yml:20]: remove this stage

File: config/samples/cert-manager_v1alpha1_kmsissuer.yaml [hunk 3]
```diff
@@ -1,9 +0,0 @@
-apiVersion: cert-manager.skyscanner.net/v1alpha1
```
Comment (Smirl) [inline config/samples/cert-manager_v1alpha1_kmsissuer.yaml:1]: revert

File: config/samples/cert-manager_v1alpha1_kmskey.yaml [hunk 4]
```diff
@@ -1,12 +0,0 @@
-apiVersion: cert-manager.skyscanner.net/v1alpha1
```
Comment (Smirl) [inline config/samples/cert-manager_v1alpha1_kmskey.yaml:1]: revert

File: config/samples/certificate.yaml [hunk 5]
```diff
@@ -1,40 +0,0 @@
----
-apiVersion: cert-manager.io/v1alpha2
```
Comment (Smirl) [inline config/samples/certificate.yaml:2]: Needs updating with the content of the README. So can be "reverted" with those new changes

File: deploy/kubernetes/kms-issuer.yaml [hunk 6]
```diff
@@ -1,515 +0,0 @@
-apiVersion: v1
-kind: Namespace
```
Comment (Smirl) [inline deploy/kubernetes/kms-issuer.yaml:2]: This is out of date with the customise. We should also add our helm chart to this repo too. So happy to delete this. Some people might be using it however. 😢

File: pkg/kmsca/kmsca.go [hunk 7]
```diff
@@ -80,7 +80,7 @@ func (ca *KMSCA) CreateKey(input *CreateKeyInput) (string, error) {
 		CustomerMasterKeySpec: aws.String(kms.CustomerMasterKeySpecRsa2048),
 	}
 	if len(input.CustomerMasterKeySpec) > 0 {
-		keyInput.CustomerMasterKeySpec = aws.String(input.CustomerMasterKeySpec)
+		keyInput.KeySpec = aws.String(input.CustomerMasterKeySpec)
```
Comment (Smirl) [inline pkg/kmsca/kmsca.go:83]: `CustomerMasterKeySpec` is depricated but at the moment we still set it on line 80. We should change it there too.

**Review by Smirl — CHANGES_REQUESTED** [review]

File: pkg/kmsca/kmsca.go (hunk 7 above)
Comment (Smirl) [reply pkg/kmsca/kmsca.go:83]: Done

File: config/rbac/cert_manager_controller_approver_clusterrolebinding.yaml [hunk 8]
```diff
@@ -1,14 +0,0 @@
-# bind the cert-manager internal approver to approve
```
[the remark this answers is not in the document]
Comment (Smirl) [reply config/rbac/cert_manager_controller_approver_clusterrolebinding.yaml]: done

File: Makefile [hunk 9]
```diff
@@ -1,10 +1,8 @@
 
 # Image URL to use all building/pushing image targets
-IMG ?= skyscanner/kms-issuer:dev
```
[the remark this answers is not in the document]
Comment (Smirl) [reply Makefile:3]: done

File: config/manager/kustomization.yaml [hunk 10]
```diff
@@ -1,10 +1,16 @@
...
+  - controller_manager_config.yaml
+  name: manager-config
 apiVersion: kustomize.config.k8s.io/v1beta1
 kind: Kustomization
 images:
 - name: controller
-  newName: skyscanner/kms-issuer
-  newTag: dev
-- name: controller=chaudyg/kms=issuer
+  newName: controller
```
[the remark this answers is not in the document]
Comment (Smirl) [reply config/manager/kustomization.yaml]: done

**Review by Smirl — APPROVED** [review]
Two unresolved comments but I'm happy to merge with it as it is right now.

File: deploy/kubernetes/kms-issuer.yaml (hunk 6 above)
Comment (maruina) [reply deploy/kubernetes/kms-issuer.yaml:2]: Yeah, I think we should keep this but we need a new PR for an automated process when we release something new

File: Makefile [hunk 11]
```diff
@@ -1,10 +1,8 @@
...
-# Produce CRDs that work back to Kubernetes 1.11 (no version conversion)
-CRD_OPTIONS ?= "crd:trivialVersions=true"
-# Use tmp kind cluster for testing
-USE_EXISTING_CLUSTER ?= "false"
```
[the remark this answers is not in the document]
Comment (maruina) [reply Makefile:7]: I'm not 100% sure what are the benefit of the kind cluster if we're not running the integration tests.

At the moment I will keep this migration to kubebuilder running the tests against `envtest` and we can raise a new issue to add proper e2e testing in kind.

What do you think?

**Review by Smirl — APPROVED** [review]

**Pull request merged**